File: externals/transportPAO/src/transport/smearing_base.f90
!
! Copyright (C) 2006 WanT Group, 2017 ERMES Group
!
! This file is distributed under the terms of the
! GNU General Public License. See the file `License\'
! in the root directory of the present distribution,
! or http://www.gnu.org/copyleft/gpl.txt .
!
!*********************************************
   MODULE smearing_base_module
!*********************************************
   USE kinds,         ONLY : dbl
   USE constants,     ONLY : ONE, TWO, PI, SQRTPI, SQRT2
   IMPLICIT NONE
   PRIVATE 
!
! Contains smearing definitions
! 
   
   PUBLIC :: smearing_func


CONTAINS

!
! subroutines
!

!**********************************************************
   FUNCTION smearing_func(x, smearing_type )
   !**********************************************************
   !
   ! evaluate the smearing function at x = ( eps - eps0 )/delta
   ! where eps, eps0 are energies and delta is the broadening parameter
   !
   ! the smearing function F is defined in such a way that
   ! 1/delta * F( x ) 
   ! 
   ! is normalized to one when integrated in energy 
   !
   ! allowed values for SMEARING_TYPE are:
   !  "lorentzian"
   !  "gaussian"
   !  "methfessel-paxton"
   !  "fermi-dirac"
   !  "marzari-vanderbilt"
   !
   IMPLICIT NONE
   !
   ! input variables
   !
   REAL(dbl)          :: smearing_func
   REAL(dbl)          :: x
   CHARACTER(*)       :: smearing_type
   !
   ! local variables
   !
   CHARACTER(13)      :: subname="smearing_func"
   REAL(dbl)          :: cost

!
!-------------------
! main body
!-------------------
!
       !
       ! dummy init
       smearing_func = ONE
       !
       SELECT CASE (TRIM(smearing_type))
       !
       CASE ( "lorentzian" )   
            !
            cost = ONE / PI
            !
            smearing_func = cost * ONE/( ONE + x**2 )  
    
       CASE ( "gaussian" )
            !
            cost = ONE / SQRTPI
            !
            smearing_func = cost * EXP( -x**2 )
            
       CASE ( "fermi-dirac", "fd" )
            !
            cost = ONE / TWO 
            !
            smearing_func = cost * ONE / ( ONE + COSH(x) )

       CASE ( "methfessel-paxton", "mp" )
            !
            ! Only N=1 is implemented at the moment
            !
            cost = ONE / SQRTPI
            !
            smearing_func = cost * EXP( -x**2 ) * ( 3.0_dbl/2.0_dbl - x**2 )

       CASE ( "marzari-vanderbilt", "mv" )
            !
            cost = ONE / SQRTPI 
            !
            smearing_func = cost * EXP( -(x- ONE/SQRT2 )**2 ) * ( TWO - SQRT2 * x )

       CASE DEFAULT
            CALL errore(subname, 'invalid smearing_type = '//TRIM(smearing_type),1)
       
       END SELECT

   END FUNCTION smearing_func

END MODULE smearing_base_module

File: src/PAOFLOW/transport/smearing/smearing_base.py
import numpy as np


def smearing_func(x: float, smearing_type: str) -> float:
    """
    Evaluate a normalized smearing function F(x) such that (1/δ)·F(x) integrates to 1.

    Parameters
    ----------
    x : float
        Dimensionless energy shift (typically (E - E₀) / δ).
    smearing_type : str
        Type of smearing:
        'lorentzian', 'gaussian', 'fermi-dirac' (or 'fd'),
        'methfessel-paxton' (or 'mp'), or 'marzari-vanderbilt' (or 'mv').

    Returns
    -------
    float
        Value of the smearing function F(x).

    Raises
    ------
    ValueError
        If the `smearing_type` is not recognized.

    Notes
    -----
    Implemented smearing types:

    - Lorentzian:
        F(x) = 1 / π(1 + x²)

    - Gaussian:
        F(x) = 1 / √π · exp(−x²)

    - Fermi-Dirac:
        F(x) = 1/2 · 1 / (1 + cosh(x))

    - Methfessel-Paxton (N=1):
        F(x) = 1 / √π · exp(−x²) · (3/2 − x²)

    - Marzari-Vanderbilt:
        F(x) = 1 / √π · exp(−(x − 1/√2)²) · (2 − √2 x)
    """
    smearing_type = smearing_type.strip().lower()

    if smearing_type == "lorentzian":
        return 1.0 / (np.pi * (1.0 + x**2))

    if smearing_type == "gaussian":
        return np.exp(-(x**2)) / np.sqrt(np.pi)

    if smearing_type in {"fermi-dirac", "fd"}:
        return 0.5 / (1.0 + np.cosh(x))

    if smearing_type in {"methfessel-paxton", "mp"}:
        return np.exp(-(x**2)) / np.sqrt(np.pi) * (1.5 - x**2)

    if smearing_type in {"marzari-vanderbilt", "mv"}:
        return (
            np.exp(-((x - 1.0 / np.sqrt(2.0)) ** 2))
            / np.sqrt(np.pi)
            * (2.0 - np.sqrt(2.0) * x)
        )

    raise ValueError(f"Unsupported smearing_type: '{smearing_type}'")
